flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java

```diff
@@ -0,0 +1,147 @@
+        finishedSplitIds.forEach(
+                splitId -> {
+                    assignedSplits.remove(splitId);
+                    knownSplits.remove(splitId);
+                });
+    }
+
+    /**
+     * Checks if split with specified id had been assigned to the reader.
+     *
+     * @param splitId split id
+     * @return {@code true} if split had been assigned, otherwise {@code false}
+     */
+    public boolean isAssigned(String splitId) {
+        return assignedSplits.contains(splitId);
+    }
+
+    /**
+     * Returns list of splits that can be assigned to readers. Does not include splits that are
+     * already assigned or finished. If shard ordering is enabled, only splits with finished parents
+     * will be returned.
+     *
+     * @return list of splits that can be assigned to readers.
+     */
+    public List<KinesisShardSplit> splitsAvailableForAssignment() {
+        return knownSplits.values().stream()
+                .filter(
+                        split -> {
+                            boolean splitIsNotAssigned = !isAssigned(split.splitId());
+                            if (preserveShardOrdering) {
+                                // Check if all parent splits were finished
+                                return splitIsNotAssigned
+                                        && verifyAllParentSplitsAreFinished(split);
+                            } else {
+                                return splitIsNotAssigned;
+                            }
+                        })
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Prepare split graph representation to store in state. Method returns only splits that are
+     * currently assigned to readers or unassigned. Finished splits are not included in the result.
+     *
+     * @param checkpointId id of the checkpoint
+     * @return list of splits with current assignment status
+     */
+    public List<KinesisShardSplitWithAssignmentStatus> snapshotState(long checkpointId) {
+        return knownSplits.values().stream()
+                .map(
+                        split -> {
+                            SplitAssignmentStatus assignmentStatus =
+                                    isAssigned(split.splitId())
+                                            ? SplitAssignmentStatus.ASSIGNED
+                                            : SplitAssignmentStatus.UNASSIGNED;
+                            return new KinesisShardSplitWithAssignmentStatus(
+                                    split, assignmentStatus);
+                        })
+                .collect(Collectors.toList());
+    }
+
+    private boolean verifyAllParentSplitsAreFinished(KinesisShardSplit split) {
+        boolean allParentsFinished = true;
+        for (String parentSplitId : split.getParentShardIds()) {
+            allParentsFinished = allParentsFinished && isFinished(parentSplitId);
+        }
+
+        return allParentsFinished;
+    }
```

Comment:
Might be good to leave a comment/tests to validate the assumptions we have made here, that this logic only works because parent splits are always by definition existing in the set of known splits.